src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -392,5 +411,100 @@ public int GetArrayRank()
 #else
         ImmutableArray<TypeName> GetGenericArguments() => _genericArguments;
 #endif
+
+#if SYSTEM_REFLECTION_METADATA
+        /// <summary>
+        /// Creates a new <see cref="TypeName" /> object that represents current simple name with provided assembly name.
+        /// </summary>
+        /// <param name="assemblyName">Assembly name.</param>
+        /// <returns>Created simple name.</returns>
+        /// <exception cref="InvalidOperationException">The current type name is not simple.</exception>
+        public TypeName WithAssemblyName(AssemblyNameInfo? assemblyName)
+        {
+            if (!IsSimple)
+            {
+                TypeNameParserHelpers.ThrowInvalidOperation_NotSimpleName(FullName);
```

Comment:
`FullName` can produce non-linear amount of work in case a long assembly name was applied to multiple generic arguments. Is it a concern? Should we omit the `FullName` in the exception message?

Reply:
The NrbfDecoder won't ever call this method for non-simple types, so I believe it's acceptable.

Reply:
I do not have a problem with it. 

I think you need to be careful in the threat-modeling about any place (including logging or error handling) that gets the full name since it may produce non-linear amount of work. Reducing the number of places that get the full name can make this easier. It is sort of defense-in-depth.

Reply:
Exception messages, particularly for this library, are an interesting beast.

The only writer I know of for NRBF is BF (and the remote procedure call parts of remoting, but it doesn't save answers).  BF won't mangle type names (more than it already does), and won't write down unsupported records (or we would have supported them).  So, where is the person getting this bad data other than from corrupted/tampered/malicious data?  They probably don't care what kind of record isn't supported, it just means the data's bad.  They probably don't care what the invalid type name was, it just means the data's bad.

Someone trying to write a new NRBF writer might end up caring, so we shouldn't discard the data, but it could easily be stored in either the property bag or a named property on a new derived exception type (if we felt it needed it), and the 0.01% who actually want the data can get it, while the 99.99% who don't care don't ever have to worry about the safety of logging exception messages.

A nicer middle ground is to capture numeric/enummy things in the message, but leave things like "arbitrary text" out.

(OK, this comment ended up lined up in S.R.MD.TypeName when I thought it was in NRBF... but pretend it's in NRBF)